src/impl_methods.rs

```diff
@@ -2832,16 +2877,20 @@ where
                 // Safe because A and B are the same type.
                 unsafe { unlimited_transmute::<B, A>(b) }
             };
-            // Delegate to mapv_into() using the wrapped closure.
-            // Convert output to a uniquely owned array of type Array<A, D>.
-            let output = self.mapv_into(f).into_owned();
-            // Change the return type from Array<A, D> to Array<B, D>.
-            // Again, safe because A and B are the same type.
-            unsafe { unlimited_transmute::<Array<A, D>, Array<B, D>>(output) }
+            // Delegate to mapv_into() to map from element type A to type A.
+            let output = self.mapv_into(f);
+            // If A and B are the same type, and if the input and output arrays
+            // have the same kind of memory representation (OwnedRepr vs
+            // OwnedArcRepr), then their memory representations should be the
+            // same type, e.g. OwnedRepr<A> == OwnedRepr<B>
+            assert!(core::any::TypeId::of::<S>() == core::any::TypeId::of::<<S as RawDataSubst<B>>::Output>());
```

Comment:
Can we change this to a `debug_assert`? My impression is that this checks a condition that we *think* the type system should be ensuring, but we'd like the assert just to be sure; in other words, it's a little mini test case. Unless there are conditions in which the types and trait bounds are satisfied but this `assert` is false?

Reply:
You are 100% correct!  The type equality should always be satisfied -- the assertions are just a little sanity check around the unsafe code.  We totally can change this to a `debug_assert!()`.  The use of `debug_assert!()` appears to be consistent with the rest of the ndarray codebase.

I will point out that the compiler nominally optimizes away the assertion if it can prove the types are equal after monomorphization.  For example, comment out line 6 of [this example](https://godbolt.org/z/hfcqxvo68) -- the assembly output does not change.

Reply:
Woah! Had no idea, but I suppose that makes sense - I do think it just expands into a conditional.

Reply:
I went ahead and changes the `assert!()` to a `debug_assert!()` as we discussed.  Is there anything else you would like to see changed for this PR to be merged?